src/lowering/desugar.ml

```diff
@@ -68,6 +68,20 @@ and exp' at note = function
     I.PrimE (I.RelPrim (!ot, o), [exp e1; exp e2])
   | S.ShowE (ot, e) ->
     I.PrimE (I.ShowPrim !ot, [exp e])
+  | S.ToCandidE (ot, es) ->
+    begin match !ot with
+    | T.Tup ts1 ->
+      I.PrimE (I.SerializePrim ts1, exps es)
+    | _ -> assert false
+    end
```

Comment:
```suggestion
  | S.ToCandidE (ot, es) ->
      I.PrimE (I.SerializePrim (List.map (typ es)), [tup es])
    end
```
Does this not work instead - also suggesting we don't need `ot`.

Reply:
I appreciate the suggestion.  I'm happy to try it out.  Thanks!

(I don't have enough intuition yet about the existing design to know what path is best overall.  Hopefully in time I will have a better sense of that for myself.)

Reply:
Just because `ot` is declared in the pattern but never used in the rhs of the match.

Reply:
This is the only place where it's used, I guess?  Okay, I'll try.  Thanks!

Reply:
Unfortunately, the suggestion does not compile, since `typ` and `tup` are not defined there.  I'm guessing that you mean `tupE` for `tup`, but I don't see any other place in that file that uses a function like `typ` (going from expressions to their types).  Rather, I had thought that was the role of `ot` in that constructor (to hold the types needed here).

Reply:
There is such a 'typ' function in `Ir_effect.ml`, and my suggestions should have been:
```
let es' = exps es in
I.PrimE (I.SerializePrim (List.map Ir_effect.typ es'), [tupE (exps es)])
```
Or similar.

Both source and ir expressions are type annotated, and you can access the type using `e.note.Note.typ` or similar (grep around for other examples).

Sorry I didn't see this earlier - just slack me if you get stuck on something.